ext/ddtrace_profiling_native_extension/collectors_cpu_and_wall_time.c

```diff
@@ -336,24 +609,45 @@ static VALUE _native_per_thread_context(DDTRACE_UNUSED VALUE _self, VALUE collec
   return per_thread_context_st_table_as_ruby_hash(state);
 }
 
-static long update_time_since_previous_sample(long *time_at_previous_sample_ns, long current_time_ns) {
+static long update_time_since_previous_sample(long *time_at_previous_sample_ns, long current_time_ns, long gc_start_time_ns) {
   // If we didn't have a time for the previous sample, we use the current one
   if (*time_at_previous_sample_ns == INVALID_TIME) *time_at_previous_sample_ns = current_time_ns;
 
-  long elapsed_time_ns = current_time_ns - *time_at_previous_sample_ns;
+  bool is_thread_doing_gc = gc_start_time_ns != INVALID_TIME;
+  long elapsed_time_ns = -1;
+
+  if (is_thread_doing_gc) {
+    bool previous_sample_was_during_gc = is_thread_doing_gc && gc_start_time_ns <= *time_at_previous_sample_ns;
+
+    if (previous_sample_was_during_gc) {
+      elapsed_time_ns = 0; // No time to account for -- any time since the last sample is going to get assigned to GC separately
+    } else {
+      elapsed_time_ns = gc_start_time_ns - *time_at_previous_sample_ns; // Capture time between previous sample and start of GC only
+    }
+  } else {
+    elapsed_time_ns = current_time_ns - *time_at_previous_sample_ns; // Capture all time since previous sample
+  }
+
+  if (elapsed_time_ns < 0) rb_raise(rb_eRuntimeError, "BUG: Unexpected negative elapsed_time_ns between samples");
+
   *time_at_previous_sample_ns = current_time_ns;
 
-  return elapsed_time_ns >= 0 ? elapsed_time_ns : 0 /* In case something really weird happened */;
+  return elapsed_time_ns;
 }
 
-static long wall_time_now_ns() {
+// Safety: This function is assumed never to raise exceptions by callers when raise_on_failure == false
+static long wall_time_now_ns(bool raise_on_failure) {
   struct timespec current_monotonic;
 
-  if (clock_gettime(CLOCK_MONOTONIC, &current_monotonic) != 0) rb_sys_fail("Failed to read CLOCK_MONOTONIC");
+  if (clock_gettime(CLOCK_MONOTONIC, &current_monotonic) != 0) {
+    if (raise_on_failure) rb_sys_fail("Failed to read CLOCK_MONOTONIC");
+    else return 0;
```

Comment:
Does zero have a special meaning? I ask because returning zero on failure seems odd in C.

Reply:
I'd think that `INVALID_TIME` would be the return here, based on the rest of the code base.

Reply:
You raise a good point. Let me try to address it in parts.

It's not only `wall_time_now_ns` that returns 0 on failure -- this was already that `cpu_time_now_ns` did.

My intention in doing so is to simplify the callers -- both sources of time are always used to compute deltas: how much cpu/wall elapsed in this period, e.g. between two calls of `cpu/wall_time_now_ns`.

Returning `INVALID_TIME` would mean that the callers would need to keep checking -- is either of the values (period started, period finished) not valid? If so, they need to be handled separately, and usually the "handling" would be setting the period to 0.

Thus, by setting it to 0, the caller can just do `finish - start` and not care about the result.

For GC tracking in particular it also simplifies a bit the error handling. We know that `wall_time_at_start_ns`/`cpu_time_at_start_ns` will be something other than `INVALID_TIME` during a GC (e.g. until `sample_after_gc` gets called), which seems a bit simpler than handling the case where we got two errors and the values are set to `INVALID_TIME` because we failed to resolve them. We would probably need to introduce a `MISSING_TIME` or similar to avoid it.

TL;DR: Using 0 seemed slightly simpler to the caller functions, so I went with it, but if you feel this is confusing I can take a stab at using something else (`INVALID_TIME` or `MISSING_TIME`?).

Reply:
`finish - start`: don't you have to still check for error if `finish` is zero but `start` is not? I see a few checks like this were introduced (`"BUG: Unexpected negative"`).

If the code actually contains fewer assertions like this by returning zero then sure. But if it's the same amount, then I suggest returning an invalid value marker. Whatever looks simpler.

Reply:
I actually did see a potential bug relating to this in a follow-up PR I'm preparing.

Let me get back to you on this one...